// vite-plugin-glsl/dist/index.js
"use strict";
/**
 * @module vite-plugin-glsl
 * @description Import shader file chunks
 * @author Ustym Ukhman <ustym.ukhman@gmail.com>
 * @version 0.2.2
 * @license MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const pluginutils_1 = require("@rollup/pluginutils");
const vite_1 = require("vite");
const loadShader_1 = tslib_1.__importDefault(require("./loadShader"));
/**
 * @const
 * @default
 * @readonly
 * @type {string}
 */
const DEFAULT_EXTENSION = 'glsl';
/**
 * @const
 * @default
 * @readonly
 * @type {readonly RegExp[]}
 */
const DEFAULT_SHADERS = Object.freeze([
    '**/*.glsl', '**/*.wgsl',
    '**/*.vert', '**/*.frag',
    '**/*.vs', '**/*.fs'
]);
/**
 * @function
 * @name glsl
 *
 * @param {PluginOptions} options Plugin config object
 *
 * @returns {Plugin} Vite plugin that converts shader code
 */
function default_1({ exclude = undefined, include = DEFAULT_SHADERS, defaultExtension = DEFAULT_EXTENSION, warnDuplicatedImports = true, compress = false } = {}) {
    let config;
    const filter = (0, pluginutils_1.createFilter)(include, exclude);
    const production = process.env.NODE_ENV === 'production';
    return {
        enforce: 'pre',
        name: 'vite-plugin-glsl',
        configResolved(resolvedConfig) {
            config = resolvedConfig;
        },
        transform(source, shader) {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                if (filter(shader))
                    return yield (0, vite_1.transformWithEsbuild)((0, loadShader_1.default)(source, shader, {
                        warnDuplicatedImports,
                        defaultExtension,
                        compress
                    }), shader, {
                        sourcemap: config.build.sourcemap && 'external',
                        minifyWhitespace: production,
                        loader: 'text',
                        format: 'esm'
                    });
            });
        }
    };
}
exports.default = default_1;
